src/mango/src/mango_selector.erl

```diff
@@ -608,35 +607,39 @@ has_required_fields_int([{[{<<"$and">>, Args}]}], RequiredFields)
 % must be covered by all children.
 has_required_fields_int([{[{<<"$or">>, Args}]} | Rest], RequiredFields) 
         when is_list(Args) ->
-    Remainder0 = lists:foldl(fun(Arg, Acc) ->
+    RequiredFields0 = lists:foldl(fun(Arg, Acc) ->
         % for each child test coverage against the full
         % set of required fields
-        Remainder = has_required_fields_int(Arg, RequiredFields),
+        ProcessedFields = has_required_fields_int(Arg, RequiredFields),
 
         % collect the remaining fields across all children
-        Acc ++ Remainder
+        Acc ++ ProcessedFields
     end, [], Args),
 
     % remove duplicate fields
-    Remainder1 = lists:usort(Remainder0),
-    has_required_fields_int(Rest, Remainder1);
+    RequiredFields1 = lists:usort(RequiredFields0),
+    has_required_fields_int(Rest, RequiredFields1);
 
 % Handle $and operator where it has peers. Required fields
 % can be covered by any child.
 has_required_fields_int([{[{<<"$and">>, Args}]} | Rest], RequiredFields) 
         when is_list(Args) ->
-    Remainder = has_required_fields_int(Args, RequiredFields),
-    has_required_fields_int(Rest, Remainder);
+    RequiredFields1 = has_required_fields_int(Args, RequiredFields),
+    has_required_fields_int(Rest, RequiredFields1);
 
 has_required_fields_int([{[{Field, Cond}]} | Rest], RequiredFields) ->
-    case Cond of
+    RequiredFields1 = case {Cond, lists:keymember(Field, 1, RequiredFields)} of
         % $exists:false is a special case - this is the only operator
-        % that explicitly does not require a field to exist
-        {[{<<"$exists">>, false}]} ->
-            has_required_fields_int(Rest, RequiredFields);
-        _ ->
-            has_required_fields_int(Rest, lists:delete(Field, RequiredFields))
-    end.
+        % that explicitly does not require a field to exist and so if a field in
+        % the index has a $exists: false in the selector for it. We cannot use this index
+        {{[{<<"$exists">>, false}]}, true} ->
+            [{exists, false}];
```

Comment:
This seems like a super subtle hack that relies on exists being an atom so that it is never matched by the rest of the selector. And assuming I'm understanding that correctly I'm rather not fond of that approach. What I'd suggest instead would be to throw here instead which is then caught in `has_requred_fields/2` where it'd return false.

Something here along the lines of `throw(exists_false_mismatch)` and then catch that and just return false.